### ISSUE TITLE
How to generate new OMEMO keys

### ISSUE THREAD

**Issue by Whisprin** [siacs/Conversations#1369]
After purging all available OMEMO fingerprints I cannot engage in a conversation with the other participant anymore.
I receive the following error message: There are no usable keys available for this contact. If you have purged any of their keys, they need to generate new ones.

Please provide a hint on how to generate new OMEMO keys. Since I couldn't find any possibility to do so: Implement a way to generate new OMEMO keys. Please forgive me if I haven't found that option.

If applicable in terms of OMEMO security maybe even allow the possibility to enable previously purged keys. (I was just testing, nothing was compromised)

Removing and re-adding the account is not a viable option since this will also generate new OTR keys.

**Comment by strb** [comment]
What is your use-case that necessitates purging and later re-enabling keys?

**Comment by Whisprin** [comment]
In my case just testing OMEMO and the clients reaction. I think the warning pop-up before purging the key is enough to prevent accidental purging. But this is just on a side note.

In my eye the real issue is on the generation of new OMEMO keys.

**Comment by strb** [comment]
Okay, then what would be a use-case that necessitates generating a new key?

**Comment by Whisprin** [comment]
If a conversation partner of mine purged my key I need to generate a new one.

**Comment by strb** [comment]
If you're building the app yourself, you can enable key regeneration through a Config.java setting. This is only there for internal testing use, and won't be incorporated into release builds. But the only way to accomplish this in the current release is to remove and re-add your account. Better yet, tell your contact to remove and re-add their account, as this will have the side effect of clearing their account's purged keys list.

Your contacts shouldn't be purging keys randomly. This feature is intended for compromised keys (e.g. stolen devices). That's why the warning dialog **clearly** states that this is an irreversible process, and that they won't be able to talk to that device ever again. If they simply want to temporarily disable a key (e.g. because they're not sure whether the device is legitimate or not), that's what the trust slider is for.

**Comment by SamWhited** [comment]
> Your contacts shouldn't be purging keys randomly. This feature is intended for compromised keys (e.g. stolen devices).

If you DO have your key compromised though and have to purge it, surely there should be a way to generate a new one without removing/readding your account? There are other consequences (losing history that has expired from the MAM archive) to removing your acount which some people may not want to trigger, but will be forced to because encryption is more important.

**Comment by strb** [comment]
@SamWhited What would that look like though? Key compromise in this contextpretty much means somebody has hacked your phone, and at that point you pretty much have to factory reset it anyway.

By the way, I'm not saying this is a "never gonna happen" feature, but so far I haven't heard or been able to come up with a compelling use-case that would necessitate it. If there's something I'm missing then by all means, enlighten me.

**Comment by SamWhited** [comment]
@strb Yah, I have no idea how it would happen, I'm just assuming that it's not impossible and that it would be inconvenient for someone who has revoked their keys to remove/readd their acount. Let's assume their phone was hacked, so they change their password, change their pin, revoke their keys. They don't need a factory reset, they just want a new key.

**Comment by Whisprin** [comment]
I can imagine two use-cases:

Someone thought that (s)he lost his phone, but later finds out that (s)he didn't. Re-adding an account will not only generate new OTR and OMEMO keys but also erase the history, discard conferences, etc.

I am doing weekly back-ups of all of my apps. If my phone is stolen I will just recover all the back-up data and manually refresh my keys, tell my contacts to distrust / purge the old ones. Allowing me to keep my history, app settings, and more.

**Comment by rigid** [comment]
maybe it would be enough, to list the consequences in the warning dialog explicitly (without being tl;dr).
i could even imaging a small "purge wizard" dialog, working around consequences (if possible), like "Your history might be lost: Do you want to make a backup now?"

**Comment by strb** [comment]
So I've thought about this some more....

@SamWhited My point with the factory reset is that after having been hacked you can't really be sure your system isn't backdoored, and then the hackers will just steal your new key as well. I mean there's probably a Norton Anti-Virus for Android or some bullshit like that, which will "quarantine" files or whatever it is that AV snakeoil does these days but...

@rigid Well losing your history is not really a consequence of purging a contact, so I don't think it makes sense warning about it in that context. The only direct consequence is that you'll never be able to talk to that device again, and this is already explicitly stated in the warning message. Deleting/Re-adding a contact (which will cause you to lose your history) is really only a workaround for this, not an intended workflow. Warning about that and/or working around those things is pretty much an endorsement that you're supposed to be using it that way (which you're not), and at that point it would make more sense to just add key regeneration as a feature.

The deeper issue here is that in normal usage, this issue just shouldn't ever happen. And it's not like adding a re-generate keys button is completely straightforward. There are a lot of UX issues with such a feature. Simply by virtue of existing, it will make users think that it's somehow a necessary feature, or that it could/should be used in regular app usage, which is not at all the case. And more importantly, the implications of re-generating keys are a lot more far reaching than purging a single key. Communicating all that without confusing the regular user is not an easy task, and I'm not sure it's worth pursuing given that you'll basically only ever need this after you've messed up somehow.

As far as I can tell, basically the only downside of deleting/re-adding an account is losing some history. And I honestly don't really see how that justifies cluttering the ManageAccountsActivity with a feature that a) will probably never be needed by 99% of the users, b) is likely to confuse those users, and c) will do a lot more damage if used accidentally than is caused by the "problem" it's trying to solve.

So in conclusion, unless someone can either convince me that there is a legitimate use-case that comes up under normal uses (so not something like "I pressed the 'do not press this' button, and now it doesn't work anymore, how can I fix this?"), or come up with a clean, non-confusing way of implementing key regeneration, I don't think this will get implemented.

**Comment by Whisprin** [comment]
@strb I think your explanation is very reasonable and I can relate with it.
So maybe it will be enough to add some kind of explanation how new keys will be generated: 

[There are no usable keys available for this contact. If you have purged any of their keys, they need to generate new ones.] Typically new keys will be automatically generated once they add their account to a supported XMPP client.

**Comment by strb** [comment]
@Whisprin Ah, very good point. I hadn't really noticed that because that got added quite recently, but that's actually a really poorly worded error message in this context, isn't it? I'll definitely change that to something that makes sense. Instructing people to generate new keys, but then not providing an option for it is a really stupid thing to do :)

**Comment by rigid** [comment]
:+1:

**Comment by volvo244** [comment]
I want to use my keys from pidgin, but i cant locate where to put them.

**Comment by strb** [comment]
@volvo244 you can't.

**Comment by NonExi** [comment]
Im running in the same issue now. 
Was trying to clean up the list of keys, becaue the other contact generated a lot of new keys in the past. I think i got a warning, that deleting keys is irreversible but i was assuming that generating new keys would be es easy as it is with OTR. I actually did not believe that warning. 
Even after reading this discussion I did not get the point in permanently disabling generating new keys and did not expect anything worse to happen.

My thoughts on this:
Why is it more critical to generate new keys in OMEMO than with OTR? 
Why could the dialog, telling that there are no keys not just enable the user to generate new keys?
If this is a very rare use case why should one expect generating new keys out of this dialog will cause more confusion than the situation as it is now? I don't understand why the user should think he has to regularly generate new keys if it would be possible to do so just in this dialog. 

The consequences in my case:
My server has no MAM archive (on purpose). So my entire history will be lost.
Re adding my account to conversations will imply the same security issues like generating new keys. But in this case for all of my contacts. I do not want all of them to check key signatures again.
In some way I feel I am punished for not using the application in the way it was intended...

**Comment by saddy001** [comment]
I removed the keys accidentally. The dialog box "please create new ones" was not helpful for me, as it doesn't tell me how. I think if Conversations aims at a larger user base, every button in the app must be fail-safe as it will be pushed, if it is available. The "non-tech user" would now remove the app as he is no longer able to chat with his contact.

**Comment by dstaubsauger** [comment]
A contact of mine and me also have the same problem, they setup their phone and afterwards they got the "no key available for this contact" when talking to me, i tried to fix it by deleting their key as well. We fell back to otr (sometimes huge delays of several hours on mobile connection) or unencrypted (we might as well use whatsapp then...) chat. The UI is confusing as it is, i did not expect the key to be permanently blacklisted and not readded upon starting an omemo conversation.

**Comment by iNPUTmice** [comment]
Since 1.14.9 the is an expert setting that lets you generate new keys
